### PR TITLE
Add Agilent 4284A

### PR DESCRIPTION
**Pull request by ConnorGCarr**
New driver for Agilent 4284A LCR meter

### CODE REVIEW

File: docs/api/instruments/agilent/agilent4284A.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+##############################
+Agilent 4284A LCR Meter
+##############################
```
Comment (BenediktBurger) [inline docs/api/instruments/agilent/agilent4284A.rst:3]: Please make the hashes as long as the text.

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        20 V in high-power mode.""",
+        validator=strict_range,
+        values=(0.005, 2),
+        dynamic=True
+    )
+
+    high_power_enabled = Instrument.measurement(
+        "OUTP:HPOW?",
+        """Get whether the high-power mode is enabled.
+        High power requires Option 001 (power amplifier / DC bias) is installed.""",
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:96]:
```suggestion
        High power requires Option 001 (power amplifier / DC bias) to be installed.""",

```

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        Maximum is 2 V by default, 40 V in high-power mode.""",
+        validator=strict_range,
+        values=(0, 2),
+        dynamic=True
+    )
+
+    bias_current = Instrument.control(
+        "BIAS:CURR?", "BIAS:CURR %g",
+        """Control the DC bias current in Amps.
+        Requires Option 001 (power amplifier / DC bias) is installed.
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:121]:
```suggestion
        Requires Option 001 (power amplifier / DC bias) to be installed.

```

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    )
+
+    auto_range_enabled = Instrument.control(
+        "FUNC:IMP:RANG:AUTO?", "FUNC:IMP:RANG:AUTO %d",
+        """Control whether the impedance auto range is enabled.""",
+        validator=strict_discrete_set,
+        values={False: 0, True: 1}
+    )
+
+    def _set_ranges(self, condition):
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:174]: maybe rename `condition` to `high_power_mode` and use booleans?

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        if self.options[0] == '0':
+            log.warning("Agilent 4284A power amplifier is not installed.")
+            return
+        self._set_ranges(1)
+        self.write("OUTP:HPOW 1")
+
+    def disable_high_power(self):
+        """Disable high power mode."""
+        self._set_ranges(0)
+        self.write("OUTP:HPOW 0")
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:209]: You can make it a (plain python) property:
~~~suggestion
    @property
    def high_power_mode_enabled(self):
        """Control whether the high power mode is enabled.
        Requires option 001 (power amplifier / DC bias) is installed.
        """
        pass  # add here the return value, if any

    @high_power_mode_enabled.setter
    def high_power_mode_enabled(self, value):
        self._set_ranges(int(value)
        self.write(f"OUTP:HPOW {int(value)}")
~~~

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+    def sweep_measurement(self, sweep_mode, sweep_values):
+        """Run list sweep measurement using sequential trigger.
+
+        :param sweep_mode (str): parameter to sweep across. Must be one of `frequency`,
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:214]: That is how you indicate the type, if you which so:
```suggestion
        :param str sweep_mode: parameter to sweep across. Must be one of `frequency`,
```

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+            "current": ("CURR", self._ac_current_values),
+            "bias_voltage": ("BIAS:VOLT", self._bias_voltage_values),
+            "bias_current": ("BIAS:CURR", self._bias_current_values)
+        }
+
+        if sweep_mode not in param_dict:
+            log.warning(
+                "Sweep mode but be one of %s, not '%s'.", list(param_dict.keys()), sweep_mode
+            )
+            return
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:233]: Why don't you raise an error?

File: pymeasure/instruments/agilent/agilent4284A.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        """Execute a bus trigger, which can be used when :meth:`~.trigger_on_bus`
+        is configured. Returns result of triggered measurement.
+        """
+        return self.values("*TRG")
+
+    def trigger_on_bus(self):
+        """Configure the trigger to detect events based on the bus trigger, which can be
+        activated by :meth:`~.trigger`.
+        """
+        self.write("TRIG:SOUR BUS")
```
Comment (BenediktBurger) [inline pymeasure/instruments/agilent/agilent4284A.py:288]: How do you configure it? It seems, that you just enable it.

File: tests/instruments/agilent/test_agilent4284A.py
Comment (BenediktBurger) [inline tests/instruments/agilent/test_agilent4284A.py:1]: Thanks for adding tests.

**Review by BenediktBurger — CHANGES_REQUESTED** [review]
Hey @ConnorGCarr,

thank you for your contribution and your patience.

I left a few comments, mainly regarding documentation.

File: pymeasure/instruments/agilent/agilent4284A.py (hunk 7 above)
Comment (ConnorGCarr) [reply pymeasure/instruments/agilent/agilent4284A.py:233]: There's no reason, I've changed it to a KeyError now.

File: pymeasure/instruments/agilent/agilent4284A.py (hunk 8 above)
Comment (ConnorGCarr) [reply pymeasure/instruments/agilent/agilent4284A.py:288]: This was something I copied over from another driver without double-checking. I've fixed this and added some trigger-related properties and methods.

**Comment by BenediktBurger** [comment]
Is it ready to be merged, from your side?

**Comment by ConnorGCarr** [comment]
Yes, all set on my end.

**Pull request merged**